### PR TITLE
KPU

### PR DESCRIPTION
**Pull request by laanwj**
Add KPU (registers from `src/drivers/include/kpu.h` in Kendryte SDK)

Note: it looks like `svd2rust` is ignoring the 64-bit register size (will have the same problem with `DMAC` and `FFT`).

### CODE REVIEW

**Comment by Disasm** [comment]
Seems like svd-parser does not support `size` at peripheral level. It's not so easy to support it properly, because this `size` field belongs to [` registerPropertiesGroup`](https://www.keil.com/pack/doc/CMSIS/SVD/html/elem_special.html#registerPropertiesGroup_gr) and can be used on different levels.

I suggest moving the `size` definition into the `register` definitions.

**Comment by laanwj** [comment]
> I suggest moving the size definition into the register definitions.

Okay, will do that, then.

File: k210.svd [hunk 1]
```diff
@@ -865,12 +865,124 @@
...
+                    <size>64</size>
+                </register>
+                <register>
+                    <name>fifo_threshold</name>
+                    <description>FIFO threshold</description>
+                    <addressOffset>0x28</addressOffset>
+                    <size>64</size>
+                    <fields>
+                        <field>
+                            <name>full</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>full_threshold</name>
```
Maybe it's better to use this to minimize confusion between `full` as flag and `threshold` as value.

File: k210.svd [hunk 2]
```diff
@@ -865,12 +865,124 @@
...
+                            <description>FIFO full threshold</description>
+                            <bitRange>[3:0]</bitRange>
+                        </field>
+                        <field>
+                            <name>empty</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>empty_threshold</name>
```

File: k210.svd [hunk 3]
```diff
@@ -865,12 +865,124 @@
...
+                    <size>64</size>
+                </register>
+                <register>
+                    <name>fifo_ctrl</name>
+                    <description>FIFO control</description>
+                    <addressOffset>0x38</addressOffset>
+                    <size>64</size>
+                    <fields>
+                        <field>
+                            <name>dma_flush</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>dma_fifo_flush</name>
```
Maybe the original naming is better here.
Seems like `_n` in `dma_fifo_flush_n` should also be reflected somehow, but headers and sources do not contain any information about these fields.

File: k210.svd [hunk 4]
```diff
@@ -865,12 +865,124 @@
...
+                            <description>Flush DMA FIFO</description>
+                            <bitRange>[0:0]</bitRange>
+                        </field>
+                        <field>
+                            <name>gs_flush</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>gs_fifo_flush</name>
```

File: k210.svd [hunk 5]
```diff
@@ -865,12 +865,124 @@
...
+                            <description>Flush GS FIFO</description>
+                            <bitRange>[1:1]</bitRange>
+                        </field>
+                        <field>
+                            <name>cfg_flush</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>cfg_fifo_flush</name>
```

File: k210.svd [hunk 6]
```diff
@@ -865,12 +865,124 @@
...
+                            <description>Flush configuration FIFO</description>
+                            <bitRange>[2:2]</bitRange>
+                        </field>
+                        <field>
+                            <name>cmd_flush</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>cmd_fifo_flush</name>
```

File: k210.svd [hunk 7]
```diff
@@ -865,12 +865,124 @@
...
+                            <description>Flush command FIFO</description>
+                            <bitRange>[3:3]</bitRange>
+                        </field>
+                        <field>
+                            <name>resp_flush</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>resp_fifo_flush</name>
```

**Review by Disasm — COMMENTED** [review]
Thanks!
Do you mind changing field names or I can merge this as is?

File: k210.svd (hunk 3 above)
Comment (laanwj) [reply k210.svd]: The register is called `fifo_ctrl` so I thought `fifo` is redundant in the field names.

**Comment by laanwj** [comment]
I'm fine with making the suggested changes

**Comment by Disasm** [comment]
Sorry, I can't commit these suggestions: there is no write permission to the corresponding branch of your repo. Could you commit them?

**Comment by laanwj** [comment]
Huh I must have accidentally turned that off for this PR, then. I've pushed a commit that should address them.

**Comment by Disasm** [comment]
Thanks!

**Pull request merged**